### PR TITLE
adds tftools.train.MultRateOptimizer, an optimizer that allows for multiple learning rates

### PR DESCRIPTION
**Pull request by jrock08**
on different variables in a graph

### CODE REVIEW

File: aiutils/tftools/train.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        self.variables = []
+
+    def add_variables(self, variables, optimizer):
+        """ Adds Variables and optimizers with different parameters.
+
+        variables (list of tf.variables): the variables to optimize wrt.
+        optimizer (tf.train.Optimizer): The corresponding optimizer.
+
+        """
+
```
Comment (BigRedT) [inline aiutils/tftools/train.py]: The function should also check that the variables have not already been added.

File: aiutils/tftools/train.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+class MultiRateOptimizer():
+    """ Class for managing a multi-rate optimization problem """
+
+    def __init__(self):
+        self.optimizers = []
+        self.variables = []
+
+    def add_variables(self, variables, optimizer):
```
Comment (BigRedT) [inline aiutils/tftools/train.py]: If the only intended purpose is to apply different learning rate to different variables but still use same optimizer such as Adam/SGD etc, then it might make sense to set the optimizer once and make `add_variables` take `variables` and 'learning_rate' as arguments instead of `optimizer`

**Comment by BigRedT** [comment]
Good script though :+1:

File: aiutils/tftools/train.py (hunk 2 above)
Comment (jrock08) [inline aiutils/tftools/train.py]: The reason I didn't do that is because optimizers have various required parameters, not just learning rate, so it would have to be optimizer_params .  Also, I could potentially see applying different optimizers to different parts of a network.  Eg.  Adam to top (new) layers and gradient descent to lower (fine-tune) layers.

I could change this such that the init takes an optional optimizer constructor, and then add_variables takes either optimizer or param_list.

I think add_variables should probably also return self so that you can write code as

`MultiRateOptimizer(tf.train.GradientDescentOptimizer).add_variables([..], params={'step_len',.1}).add_variables([...], params={'step_len', .01})`

File: aiutils/tftools/train.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        variables (list of tf.variables): the variables to optimize wrt.
+
+        Either:
+        optimizer (tf.train.Optimizer): The corresponding optimizer.
+        Or:
+        learning_rate (float): A learning rate to pass to the default_optimizer
+        other_params (dict): A dictionary of param_name, value to pass the the default optimizer
+        """
+
+        print 'test'
```
Comment (BigRedT) [inline aiutils/tftools/train.py]: delete print

File: aiutils/tftools/train.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        chck_vars = self.check_variables(variables)
+        if len(chck_vars) != 0:
```
Comment (BigRedT) [inline aiutils/tftools/train.py]: can just do `if chck_vars:`

File: aiutils/tftools/train.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            raise ValueError('Expected all new variables, got overlap', *
+                             [v.name for v in chck_vars])
+        assert (len(self.check_variables(variables)) == 0)
```
Comment (BigRedT) [inline aiutils/tftools/train.py]: duplicate `assert`?

File: aiutils/tftools/train.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        self.variables.append(variables)
+
+        if (optimizer is not None):
```
Comment (BigRedT) [inline aiutils/tftools/train.py]: `if optimizer:`

File: aiutils/tftools/train.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            self.optimizers.append(optimizer)
+        else:
+            if self.default_optimizer is None:
+                raise ValueError(
+                    'default_optimizer is None',
+                    'When optimizer is not passed to add_variables, expect default_optimizer to be not None')
```
Comment (BigRedT) [inline aiutils/tftools/train.py]: Rephrase error msg. Maybe -
`default_optimizer can not be None if optimizer is not passed to add_variables`

**Comment by BigRedT** [comment]
Feel free to merge after fixing the minor comments
